### PR TITLE
chore: route builder refactor

### LINKED ISSUE #27840: [Task]: Refactor RouteBuilder to support Packages

### PR DESCRIPTION
**Pull request by ashit-rath**
## Description
This PR aims to achieve 2 things
1. Make route building independent of explicitly passing `pageId` as param when the route is generated against the current page navigation.
2. Add extensible points to extend basePath generation in EE.

THIS PR DOES NOT CHANGE ROUTE GENERATION. 

Changes: 
In `app/client/src/ce/entities/URLRedirect/URLAssembly.ts`
1. Moves the logic of `generateBasePath` way to specific method called `generateBasePathForApps` and the generateBasePath is available to extend and switch between a different base path generation logic in EE.
2. Adds a new member variable called `currentPageId`. This `currentPageId` would help generating basePath without explicitly passing `pageId` to the build method. If a `pageId` is passed it would be overridden in the `resolveEntityId` logic.
3. Added `resolveEntityId` method to resolve the entity (pageId) based on the params passed and the `currentPageId`. This method also acts as an extension point for extending the logic to any other resolution logic similar to `generateBasePath`

In `app/client/src/pages/AppViewer/index.tsx` and `app/client/src/pages/Editor/index.tsx`
The `currentPageId` is set using the `urlBuilder.setCurrentPageId` when the component mounts or page changes and unset when the component unmounts.

#### PR fixes following issue(s)
Fixes #27840 

#### Type of change
- Chore (housekeeping or task changes that don't impact user perception)

- [x] Manual
- [ ] JUnit
- [x] Jest
- [ ] Cypress
>
>
#### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

#### QA activity:
- [ ] [Speedbreak features](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#speedbreakers-) have been covered
- [ ] Test plan covers all impacted features and [areas of interest](https://github.com/appsmithorg/TestSmith/wiki/Guidelines-for-test-plans#areas-of-interest-)
- [ ] Test plan has been peer reviewed by project stakeholders and other QA members
- [ ] Manually tested functionality on DP
- [ ] We had an implementation alignment call with stakeholders post QA Round 2
- [ ] Cypress test cases have been added and approved by SDET/manual QA
- [ ] Added `Test Plan Approved` label after Cypress tests were reviewed
- [ ] Added `Test Plan Approved` label after JUnit tests were reviewed

### CODE REVIEW

File: app/client/src/RouteParamsMiddleware.ts [hunk 1]
```diff
@@ -8,8 +8,8 @@ import type { UpdatePageResponse } from "api/PageApi";
 import type {
   ApplicationURLParams,
   PageURLParams,
-} from "entities/URLRedirect/URLAssembly";
```
Comment (arunvjn) [inline app/client/src/RouteParamsMiddleware.ts:11]: We might need to split store.ts if we planning to cache modules info into the extends URLBuilder class

**Review by arunvjn — APPROVED** [review]
Changes look good!

**Review by ankitakinger — DISMISSED** [review]

**Review by ankitakinger — DISMISSED** [review]

**Review by ankitakinger — APPROVED** [review]
LGTM

**Pull request merged**